x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/QueryRule.java

```diff
@@ -0,0 +1,258 @@
+
+    private final String id;
+    private final QueryRuleType type;
+    private final List<QueryRuleCriteria> criteria;
+    private final Map<String, Object> actions;
+
+    public enum QueryRuleType {
+        PINNED;
+
+        public static QueryRuleType queryRuleType(String type) {
+            for (QueryRuleType queryRuleType : QueryRuleType.values()) {
+                if (queryRuleType.name().equalsIgnoreCase(type)) {
+                    return queryRuleType;
+                }
+            }
+            throw new IllegalArgumentException("Unknown QueryRuleType: " + type);
+        }
+    }
+
+    /**
+     * Public constructor.
+     *
+     * @param id                        The unique identifier associated with this query rule
+     * @param type                      The {@link QueryRuleType} of this rule
+     * @param criteria                  The {@link QueryRuleCriteria} required for a query to match this rule
+     * @param actions                   The actions that should be taken if this rule is matched, dependent on the type of rule
+     */
+    public QueryRule(String id, QueryRuleType type, List<QueryRuleCriteria> criteria, Map<String, Object> actions) {
+        if (Strings.isNullOrEmpty(id)) {
+            throw new IllegalArgumentException("Query rule id cannot be null or blank");
+        }
+        this.id = id;
+
+        Objects.requireNonNull(type, "Query rule type cannot be null");
+        this.type = type;
+
+        Objects.requireNonNull(criteria, "Query rule criteria cannot be null");
+        if (criteria.isEmpty()) {
+            throw new IllegalArgumentException("Query rule criteria cannot be empty");
+        }
+        this.criteria = criteria;
+
+        Objects.requireNonNull(actions, "Query rule actions cannot be null");
+        if (actions.isEmpty()) {
+            throw new IllegalArgumentException("Query rule actions cannot be empty");
+        }
+        this.actions = actions;
+
+        validate();
+    }
+
+    public QueryRule(StreamInput in) throws IOException {
+        this.id = in.readString();
+        this.type = QueryRuleType.queryRuleType(in.readString());
+        this.criteria = in.readList(QueryRuleCriteria::new);
+        this.actions = in.readMap();
+
+        validate();
+    }
+
+    private void validate() {
+        if (type == QueryRuleType.PINNED) {
+            if (actions.containsKey("ids") == false && actions.containsKey("docs") == false) {
+                throw new ElasticsearchParseException("Pinned Query rule actions must contain either ids or docs");
+            }
+        } else {
+            throw new IllegalArgumentException("Unsupported QueryRuleType: " + type);
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(id);
+        out.writeString(type.toString());
+        out.writeList(criteria);
+        out.writeGenericMap(actions);
+    }
+
+    @SuppressWarnings("unchecked")
```

Comment:
Do we have to suppress warnings at the method level?

Reply:
This is personal preference, to avoid multiple `@SuppressWarnings` in a single method. It felt cleaner. I can change it if you prefer.

Reply:
Makes sense, there will be multiple ones. Thanks!